### PR TITLE
executor, types: fix char column not compatible with mysql

### PR DESCRIPTION
**Pull request by zhuozl**
### What problem does this PR solve? 

for #10392 

### What is changed and how it works?

add condition in  types/datum.go

### Check List 

Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

```
create table vctt (v varchar(4), c char(4))
mysql> insert into vctt values ("ab  ", "  ab   ");
Query OK, 1 row affected (0.04 sec)
mysql> select * from vctt;
+------+------+
| v    | c    |
+------+------+
| ab   | NULL |
| ab   | NULL |
| NULL | ab   |
| NULL | ab   |
| ab   | ab   |
| ab   |   ab |
| ab   |   ab |
| ab   | ab   |
+------+------+
8 rows in set (0.00 sec)

```

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -798,6 +798,9 @@ func (d *Datum) convertToString(sc *stmtctx.StatementContext, target *FieldType)
 		s = strconv.FormatFloat(d.GetFloat64(), 'f', -1, 64)
 	case KindString, KindBytes:
 		s = d.GetString()
+		if target.Tp == mysql.TypeString {
+			s=strings.TrimRight(s, " ")
```
Comment (lonng) [inline types/datum.go]: I am sure `TrimRight` is not the correct behavior.
```
mysql> insert into vctt values ("ab  ", "  ab   ");
mysql> select concat('(', v, ')'), concat('(', c, ')') from vctt;
+-------------------------------------------+
| concat('(', v, ')') | concat('(', c, ')') |
+-------------------------------------------+
| (ab  )              | (ab)                |
+-------------------------------------------+
3 row in set (0.002 sec)
```

Comment (lonng) [reply types/datum.go]: Please format the code before commit, thanks.

Comment (zhuozl) [reply types/datum.go]:
```
TiDB:
mysql> select concat('(', v, ')'), concat('(', c, ')') from vctt;
+---------------------+---------------------+
| concat('(', v, ')') | concat('(', c, ')') |
+---------------------+---------------------+
| (ab  )              | (  ab)              |
| (ab  )              | (  ab)              |
+---------------------+---------------------+
2 rows in set (0.00 sec)

mysql:
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.16    |
+-----------+
1 row in set (0.00 sec)

mysql> select concat('(', v, ')'), concat('(', c, ')') from vctt;
+---------------------+---------------------+
| concat('(', v, ')') | concat('(', c, ')') |
+---------------------+---------------------+
| (ab  )              | (  ab)              |
+---------------------+---------------------+
1 row in set (0.00 sec)

```

File: types/datum.go [hunk 2]
```diff
@@ -798,6 +798,9 @@ func (d *Datum) convertToString(sc *stmtctx.StatementContext, target *FieldType)
 		s = strconv.FormatFloat(d.GetFloat64(), 'f', -1, 64)
 	case KindString, KindBytes:
 		s = d.GetString()
+		if target.Tp == mysql.TypeString {
+			s = strings.TrimRight(s, " ")
```
Comment (XuHuaiyu) [inline types/datum.go]: Should we consider other blank characters here? `\n`, `\t`,etc

File: types/datum.go (hunk 1 above)
Comment (zhuozl) [reply types/datum.go]: I did not find specific trailing spaces in "https://dev.mysql.com/doc/refman/8.0/en/char.html".
I tried the following situations and the results are the same.
```
insert into vctt values ("ab  ", " a\a\a\a\a");
insert into vctt values ("ab  ", " a\b\b\b\b");
insert into vctt values ("ab  ", " a\f\f\f\f");
insert into vctt values ("ab  ", " a\v\v\v\v");

insert into vctt values ("ab  ", " a\n\n\n\n");
insert into vctt values ("ab  ", " a\r\r\r\r");
insert into vctt values ("ab  ", " a\t\t\t\t");
insert into vctt values ("ab  ", " a        ");

insert into vctt values ("ab  ", " a\\\\\\\\");
insert into vctt values ("ab  ", " a\'\'\'\'");
insert into vctt values ("ab  ", " a\"\"\"\"");
insert into vctt values ("ab  ", " a\?\?\?\?");
insert into vctt values ("ab  ", " a\0\0\0\0");
```

File: types/datum.go (hunk 2 above)
Comment (zhuozl) [reply types/datum.go]: I did not find specific trailing spaces in "https://dev.mysql.com/doc/refman/8.0/en/char.html".
I tried the following situations and the results are the same.
```
insert into vctt values ("ab  ", " a\a\a\a\a");
insert into vctt values ("ab  ", " a\b\b\b\b");
insert into vctt values ("ab  ", " a\f\f\f\f");
insert into vctt values ("ab  ", " a\v\v\v\v");

insert into vctt values ("ab  ", " a\n\n\n\n");
insert into vctt values ("ab  ", " a\r\r\r\r");
insert into vctt values ("ab  ", " a\t\t\t\t");
insert into vctt values ("ab  ", " a        ");

insert into vctt values ("ab  ", " a\\\\\\\\");
insert into vctt values ("ab  ", " a\'\'\'\'");
insert into vctt values ("ab  ", " a\"\"\"\"");
insert into vctt values ("ab  ", " a\?\?\?\?");
insert into vctt values ("ab  ", " a\0\0\0\0");
```

File: types/datum.go [hunk 3]
```diff
@@ -848,7 +848,17 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *stmtctx.StatementCon
 					}
 					runeCount++
 				}
-				err = ErrDataTooLong.GenWithStack("Data Too Long, field len %d, data len %d", flen, characterLen)
+
+				truncateTrailingSpacesLen := math.MaxInt32
+				if tp.Tp == mysql.TypeString {
```
Comment (zz-jason) [inline types/datum.go]: we should also consider binary strings, consider this case:
```sql
drop table if exists t;
create table t(a char(2), b char(2) binary, c binary(2));
insert into t values("aa ", null, null);
insert into t values(null, "aa ", null);
insert into t values(null, null, "aa ");
```

In MySQL, only the last insert statement fails:
```sql
MySQL(root@localhost:test) > drop table if exists t;
Query OK, 0 rows affected (0.08 sec)

MySQL(root@localhost:test) > create table t(a char(2), b char(2) binary, c binary(2));
Query OK, 0 rows affected (0.11 sec)

MySQL(root@localhost:test) > insert into t values("aa ", null, null);
Query OK, 1 row affected (0.02 sec)

MySQL(root@localhost:test) > insert into t values(null, "aa ", null);
Query OK, 1 row affected (0.04 sec)

MySQL(root@localhost:test) > insert into t values(null, null, "aa ");
ERROR 1406 (22001): Data too long for column 'c' at row 1
MySQL(root@localhost:test) > select version();
+-----------+
| version() |
+-----------+
| 8.0.11    |
+-----------+
1 row in set (0.00 sec)
```

Comment (zhuozl) [reply types/datum.go]: I tested the above statements and the results are the same.

```
mysql> create table t(a char(2), b char(2) binary, c binary(2));
Query OK, 0 rows affected (0.54 sec)

mysql> insert into t values("aa ", null, null);
Query OK, 1 row affected (0.04 sec)

mysql> insert into t values(null, "aa ", null);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t values(null, null, "aa ");
ERROR 1406 (22001): Data too long for column 'c' at row 1
mysql> select version();
+------------------+
| version()        |
+------------------+
| 5.7.25-TiDB-None |
+------------------+
1 row in set (0.00 sec)

```

Comment (zhuozl) [reply types/datum.go]: @zz-jason  Is it Right?

File: executor/seqtest/seq_executor_test.go [hunk 4]
```diff
@@ -458,6 +458,13 @@ func (s *seqTestSuite) TestShow(c *C) {
 	result = tk.MustQuery(`show plugins like '%'`)
 	result.Check(result.Rows())
 
+	//for issue #10392
+	testSQL = `drop table if exists vctt`
+	tk.MustExec(testSQL)
+	testSQL = `create table vctt (v varchar(4), c char(4))`
+	tk.MustExec(testSQL)
+	tk.MustExec(`insert into vctt values ("ab  ", "ab   ")`)
```
Comment (Deardrops) [inline executor/seqtest/seq_executor_test.go]: Don't forget to check the correctness of the results.
```suggestion
	tk.MustExec(`insert into vctt values ("ab  ", "ab   ")`)
	tk.MustQuery("select * from vctt").Check(testkit.Rows("ab ab"))
```

**Comment by lonng** [comment]
Ping @zhuozl

**Pull request closed without merging**